### PR TITLE
feat: keychain

### PR DESCRIPTION
**Pull request by vasco-santos**
This PR is part of the [PeerStore improvements EPIC](https://github.com/libp2p/js-libp2p/issues/582) and tackles part of the milestone 4 - [KeyChain](https://github.com/libp2p/js-libp2p/issues/591).

Points to note:
- The entire libp2p-keychain history has been added
- Additional commits were made to move keychain under `./src/keychain` and its tests under `./test/keychain`.
- Some of the unneeded files were removed (like LICENSE) since libp2p already has a version of that file
- libp2p-keychain is now created on `Libp2p.create` and follows the same approach as `js-ipfs` did https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs/src/core/components/init.js#L178-L257 (This will not be needed anymore in ipfs, you can also see this temporary commit with the new usage: https://github.com/ipfs/js-ipfs/commit/1a342c5dbc481596c214348f044d0ac09164d3fe)

## **Note on Merging**: 

Right now the plan (obtained from `libp2p-keychain` PR) is to:

- squash the merge to minimize the commit pollution
- update the readme to point to the original repo
- append the Co-Author list to the squashed commit to maintain author credit
  - This co author list was generated from the libp2p-keychain package.json which is auto generated from the commit history there.

```
Co-authored-by: Alan Shaw <alan.shaw@protocol.ai>
Co-authored-by: Alberto Elias <hi@albertoelias.me>
Co-authored-by: Alex Potsides <alex@achingbrain.net>
Co-authored-by: David Dias <daviddias.p@gmail.com>
Co-authored-by: Hugo Dias <mail@hugodias.me>
Co-authored-by: Jacob Heun <jacobheun@gmail.com>
Co-authored-by: Maciej Krüger <mkg20001@gmail.com>
Co-authored-by: Masahiro Saito <camelmasa@gmail.com>
Co-authored-by: Richard Schneider <makaretu@gmail.com>
Co-authored-by: Vasco Santos <vasco.santos@moxy.studio>
Co-authored-by: Vasco Santos <vasco.santos@ua.pt>
Co-authored-by: Victor Bjelkholm <victorbjelkholm@gmail.com>
```

This replaces #525 because the other PR was super outdated and complex to rebase at this point.

### CODE REVIEW

File: doc/API.md [hunk 1]
```diff
@@ -470,6 +482,65 @@ const listenMa = libp2p.transportManager.getAddrs()
...
+|------|-------------|
+| `Array<Multiaddr>` | listening multiaddrs |
+
+#### Example
+
+```js
+// ...
+const listenMa = libp2p.transportManager.getAddrs()
+// [ <Multiaddr 047f00000106f9ba - /ip4/127.0.0.1/tcp/63930> ]
+
```
Comment (jacobheun) [inline doc/API.md:543]: This whole added block already exists, this is adding duplicate data.

Comment (jacobheun) [reply doc/API.md:543]: Removing this should also fix the formatting below as it isn't properly closing the code blocks

File: doc/API.md [hunk 2]
```diff
@@ -1254,6 +1325,278 @@ libp2p.connectionManager.size
...
+| Type | Description |
+|------|-------------|
+| `Promise<Buffer>` | Decrypted data. |
+
+#### Example
+
+```js
+const keyInfo = await libp2p.keychain.createKey('keyTest', 'rsa', 4096)
+const enc = await libp2p.keychain.cms.encrypt('keyTest', Buffer.from('data'))
```
Comment (jacobheun) [inline doc/API.md:1597]: This is an encrypt example, but this section is decrypt

File: src/index.js [hunk 3]
```diff
@@ -44,8 +47,9 @@ class Libp2p extends EventEmitter {
     // and add default values where appropriate
     this._options = validateConfig(_options)
 
-    this.peerId = this._options.peerId
     this.datastore = this._options.datastore
+    this.keychain = this._options.keychain
```
Comment (jacobheun) [inline src/index.js:51]: `_options.keychain` is an object, not a KeyChain, this will break

File: src/index.js [hunk 4]
```diff
@@ -168,6 +172,9 @@ class Libp2p extends EventEmitter {
     this.peerRouting = peerRouting(this)
     this.contentRouting = contentRouting(this)
 
+    // Keychain
+    this.keychain = this._options._keychain || new NoKeychain()
```
Comment (jacobheun) [inline src/index.js:176]: Why `NoKeychain`? It's just throwing errors whenever you try to use it. Just leave it undefined and let type errors happen.

File: src/index.js [hunk 5]
```diff
@@ -542,13 +548,40 @@ class Libp2p extends EventEmitter {
...
+    log('keychain constructed')
+
+    // Import the private key as 'self', if needed.
+    try {
+      await keychain.findByName('self')
+    } catch (err) {
+      await keychain.importPeer('self', peerId)
+    }
+
+    options._keychain = keychain
```
Comment (jacobheun) [inline src/index.js:582]: Having keychain and _keychain on options is confusing. Why not move the keychain creation in to the libp2p constructor? We can create libp2p then do the keychain await stuff before returning the libp2p instance.

**Comment by vasco-santos** [comment]
Closed in favour of #633 and #634

**Pull request closed without merging**